src/com/ms/silverking/cloud/ring/RingRegion.java

```diff
@@ -13,210 +13,208 @@
-        
-        assert subRegionSize > 0 && subRegionSize < getSize(); 
-        subRegionEnd = LongRingspace.add(start, subRegionSize - 1);
-        return new RingRegion(start, subRegionEnd);
+  private final long start;
+  private final long end;
+
+  public static final RingRegion allRingspace = new RingRegion(LongRingspace.start, LongRingspace.end);
+
+  // used for ordering RingRegions
+  public static final Comparator<RingRegion> sizeComparator = new RingRegionSizeComparator();
+  public static final Comparator<RingRegion> positionComparator = new RingRegionPositionComparator();
+  public static final Comparator<RingRegion> sizePositionComparator = new RingRegionSizePositionComparator();
+
+  public RingRegion(long start, long end) {
+    LongRingspace.ensureInRingspace(start);
+    LongRingspace.ensureInRingspace(end);
+    if (getSize(start, end) == LongRingspace.size) {
+      // normalize any region that consumes the entire ringspace
+      start = LongRingspace.start;
+      end = LongRingspace.end;
+    }
+    this.start = start;
+    this.end = end;
+  }
+
+  public long getStart() {
+    return start;
+  }
+
+  public long getEnd() {
+    return end;
+  }
+
+  private static long getSize(long start, long end) {
+    return LongRingspace.clockwiseDistance(start, end) + 1;
+  }
+
+  public long getSize() {
+    return getSize(start, end);
+  }
+
+  public boolean before(long p0, long p1) {
+    long d0;
+    long d1;
+
+    d0 = LongRingspace.clockwiseDistance(start, p0);
+    d1 = LongRingspace.clockwiseDistance(start, p1);
+    return d0 < d1;
+  }
+
+  public boolean after(long p0, long p1) {
+    long d0;
+    long d1;
+
+    d0 = LongRingspace.clockwiseDistance(start, p0);
+    d1 = LongRingspace.clockwiseDistance(start, p1);
+    return d0 > d1;
+  }
+
+  public BigDecimal getSizeBD() {
+    return new BigDecimal(getSize(), LongRingspace.mathContext);
+  }
+
+  public double getRingspaceFraction() {
+    return getSizeBD().divide(LongRingspace.bdSize, LongRingspace.mathContext).doubleValue();
+  }
+
+  public static void ensureIdentical(RingRegion r0, RingRegion r1) {
+    if (!r0.equals(r1)) {
+      System.err.println(r0);
+      System.err.println(r1);
+      throw new RuntimeException("Regions not identical");
     }
-    
-    public RingRegion partitionEnd(long subRegionSize) {
-        long    subRegionStart;
-    
-        assert subRegionSize > 0 && subRegionSize < getSize(); 
-        subRegionStart = LongRingspace.subtract(end, subRegionSize - 1);
-        return new RingRegion(subRegionStart, end);
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    RingRegion oRegion;
+
+    oRegion = (RingRegion) other;
+    return this.start == oRegion.start && this.end == oRegion.end;
+  }
+
+  @Override
+  public int hashCode() {
+    return (int) start ^ (int) end;
+  }
+
+  public static RingRegion parseZKString(String s) {
+    try {
+      int index;
+
+      index = s.indexOf(':');
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

If the indexOf() search on line 105 does not find the substring and returns -1, the call to substring() on line 106 will throw IndexOutOfBoundsException. Note that the method is publicly accessible to external callers who might not be aware of the constraint on string s.